polymer_brushes/SCF.py

```diff
@@ -0,0 +1,899 @@
+## This code was adapted from that of Paul Kienzle (NIST) and Richard Sheridan.
+## The original code was public domain.
+
+## The original code was designed for use with Refl1D.
+## Isaac Gresham modified it for compatability with refnx.
+
+## Further optimisations are required for nSCFT_VFP to be
+## compatible with
+
+from collections import OrderedDict
+import numpy as np
+from numpy.core.multiarray import correlate as old_correlate
+
+from scipy.interpolate import PchipInterpolator as Pchip
+from scipy.integrate import simps
+
+from refnx.reflect import Structure, Component, SLD, Slab
+from refnx.analysis import Parameters, Parameter, possibly_create_parameter
+
+LAMBDA_1 = 1.0 / 6.0  # always assume cubic lattice (1/6) for now
+LAMBDA_0 = 1.0 - 2.0 * LAMBDA_1
+# Use reverse order for LAMBDA_ARRAY if it is asymmetric since we are using
+# it with correlate().
+LAMBDA_ARRAY = np.array([LAMBDA_1, LAMBDA_0, LAMBDA_1])
+MINLAT = 25
+MINBULK = 5
+SQRT_PI = np.sqrt(np.pi)
+
+_SZdist_dict = OrderedDict()
+
+
+EPS = np.finfo(float).eps
+
+
+class nSCFT_VFP(Component):
+    def __init__(
+        self,
+        adsorbed_amount,
+        lattice_size,
+        polymerMW,
+        latticeMW,
+        chi,
+        chi_s,
+        pdi,
+        polymer_sld,
+        rough=1,
+        name="",
+        microslab_max_thickness=1,
+        dist_model=None,
+        clear_cache=False,
+    ):
+        """
+        Parameters
+        ----------
+        Adsorbed Amount : Parameter or float
+            The integral of the volume fraction profile, equivalent to the dry thickness
+
+        polymer_sld : SLD object or float
+            The scattering length density of the polymer component
+
+        chi : float
+            Flory-Huggins solvent interaction parameter
+
+        chi_s :
+            surface interaction parameter
+
+        l_lat :
+            real length per lattice site
+
+        mn : float
+            Number average molecular weight
+
+        m_lat : float
+            real mass per lattice segment
+
+        pdi : float
+            Dispersity (Polydispersity index). float >= 1. By default,
+            calculated using the schulz-zimm distribution
+
+        name : string
+            The name of the structural component. Useful for later introspeciton.
+
+        microslab_max_thickness : float
+            Maximum slab thickness when approximating the profile as a series of slabs.
+
+        rough : float
+            the (gaussian) roughness between slabs in the slab approximation of the profile
+
+        dist_model : function or None
```

Comment:
What is dist_model, is it an rv_continuous, or a general callable? What is the call signature that the function should take?